projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/control/LLVMDispatchBasicBlockNode.java

```diff
@@ -99,36 +100,26 @@ public Object executeGeneric(VirtualFrame frame) {
         outer: while (basicBlockIndex != LLVMBasicBlockNode.RETURN_FROM_FUNCTION) {
             CompilerAsserts.partialEvaluationConstant(basicBlockIndex);
 
-            // TODO restructure, to have LoopNode fit into LLVMControlFlowNode structure
-            if(bodyNodes[basicBlockIndex] instanceof LLVMLoopNode) {
-                LLVMLoopNode loop = (LLVMLoopNode) bodyNodes[basicBlockIndex];
-                loop.execute(frame);
-                int successorBasicBlockIndex = FrameUtil.getIntSafe(frame, loopSuccessorSlot);
+            LLVMControlFlowNode controlFlowNode;
+            LLVMBasicBlockNode bb = null;
+            /*
+             * Distinction between low level basic blocks and high level control flow
+             * structures, like LLVMLoopNode
+             */
+            if(bodyNodes[basicBlockIndex] instanceof LLVMBasicBlockNode) {
```

Comment:
This check will probably have a noticeable impact on performance. It should not be necessary either.